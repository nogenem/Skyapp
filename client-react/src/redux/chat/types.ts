import { TMessageType } from '~/constants/message_types';
import { TUserStatus } from '~/constants/user_status';

interface IOtherUser {
  _id: string;
  nickname: string;
  thoughts: string;
  status: TUserStatus;
  online: boolean;
  channel_id?: string;
}

interface IOtherUsers {
  [_id: string]: IOtherUser;
}

interface IMember {
  user_id: string;
  is_adm: boolean;
  last_seen: Date;
}

interface IChannel {
  _id: string;
  name: string;
  is_group: boolean;
  members: IMember[];
  other_member_idx?: number;
  unread_msgs: number;
  lastMessage?: IMessage;
}

interface IChannels {
  [_id: string]: IChannel;
}

interface IAttachment {
  originalName: string;
  size: number;
  path: string;
  mimeType: string;
  imageDimensions?: { width: number; height: number };
}

interface IMessage {
  _id: string;
  channel_id: string;
  from_id?: string;
  body: string | IAttachment;
  type: TMessageType;
  createdAt: Date;
  updatedAt: Date;
}

interface IInitialData {
  users: IOtherUsers;
  channels: IChannels;
}

interface IActiveChannelInfo {
  _id: string;
  messages: IMessage[];
  totalMessages: number;
  queue: IMessage[];
}

interface INewGroupCredentials {
  name: string;
  members: string[];
  admins: string[];
}

interface IUpdateGroupCredentials {
  channel_id: string;
  name: string;
  members: string[];
  admins: string[];
}

interface ILeaveGroupCredentials {
  channel_id: string;
}

interface IFetchMessagesCredentials {
  channel_id: string;
  offset: number;
  limit?: number;
  sort?: string;
}

interface ISendMessageCredentials {
  channel_id: string;
  body: string;
}

type TChatState = {
  users: IOtherUsers;
  channels: IChannels;
  activeChannelInfo?: IActiveChannelInfo;
};

enum EChatActions {
  SET_INITIAL_DATA = '@chat/SET_INITIAL_DATA',
  SET_USER_ONLINE = '@chat/SET_USER_ONLINE',
  ADD_OR_UPDATE_CHANNEL = '@chat/ADD_OR_UPDATE_CHANNEL',
  SET_ACTIVE_CHANNEL = '@chat/SET_ACTIVE_CHANNEL',
  REMOVE_CHANNEL = '@chat/REMOVE_CHANNEL',
  ADD_NEW_USER = '@chat/ADD_NEW_USER',
  ADD_MESSAGES = '@chat/ADD_MESSAGES',
  SET_LATEST_MESSAGE = '@chat/SET_LATEST_MESSAGE',
  ADD_TO_MESSAGES_QUEUE = '@chat/ADD_TO_MESSAGES_QUEUE',
  REMOVE_FROM_MESSAGES_QUEUE = '@chat/REMOVE_FROM_MESSAGES_QUEUE',
  SET_LAST_SEEN = '@chat/SET_LAST_SEEN',
  SET_USER_STATUS = '@chat/SET_USER_STATUS',
  SET_USER_THOUGHTS = '@chat/SET_USER_THOUGHTS',
}

interface IChatActionType<T, P> {
  type: T;
  payload: P;
}

type TChatAction =
  | IChatActionType<typeof EChatActions.SET_INITIAL_DATA, IInitialData>
  | IChatActionType<
      typeof EChatActions.SET_USER_ONLINE,
      { _id: string; value: boolean }
    >
  | IChatActionType<typeof EChatActions.ADD_OR_UPDATE_CHANNEL, IChannel>
  | IChatActionType<
      typeof EChatActions.SET_ACTIVE_CHANNEL,
      { _id: string | undefined }
    >
  | IChatActionType<typeof EChatActions.REMOVE_CHANNEL, { channelId: string }>
  | IChatActionType<typeof EChatActions.ADD_NEW_USER, IOtherUser>
  | IChatActionType<
      typeof EChatActions.ADD_MESSAGES,
      { messages: IMessage[]; totalMessages: number; atTop: boolean }
    >
  | IChatActionType<typeof EChatActions.SET_LATEST_MESSAGE, IMessage>
  | IChatActionType<typeof EChatActions.ADD_TO_MESSAGES_QUEUE, IMessage>
  | IChatActionType<typeof EChatActions.REMOVE_FROM_MESSAGES_QUEUE, IMessage>
  | IChatActionType<
      typeof EChatActions.SET_LAST_SEEN,
      {
        channel_id: string;
        user_id: string;
        last_seen: string;
      }
    >
  | IChatActionType<
      typeof EChatActions.SET_USER_STATUS,
      {
        user_id: string;
        newStatus: TUserStatus;
      }
    >
  | IChatActionType<
      typeof EChatActions.SET_USER_THOUGHTS,
      {
        user_id: string;
        newThoughts: string;
      }
    >;

export type {
  TChatState,
  TChatAction,
  IInitialData,
  IOtherUsers,
  IOtherUser,
  IChannel,
  IChannels,
  IMember,
  IMessage,
  IAttachment,
  INewGroupCredentials,
  IUpdateGroupCredentials,
  ILeaveGroupCredentials,
  IFetchMessagesCredentials,
  ISendMessageCredentials,
};
export { EChatActions };
