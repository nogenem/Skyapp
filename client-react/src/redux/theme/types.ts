type TThemeMode = 'dark' | 'light';

type TThemeState = {
  mode?: TThemeMode;
};

enum EThemeActions {
  SWITCH_MODE = '@theme/SWITCH_MODE',
}

interface IThemeActionType<T, P> {
  type: T;
  payload: P;
}

type TThemeAction = IThemeActionType<
  typeof EThemeActions.SWITCH_MODE,
  TThemeMode
>;

export type { TThemeMode, TThemeState, TThemeAction };
export { EThemeActions };
