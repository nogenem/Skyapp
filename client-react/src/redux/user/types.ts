interface User {
  _id: string;
  token?: string;
}

type UserState = User;

enum EUserActions {
  SIGNED_IN = '@user/SIGNED_IN',
  SIGNED_OUT = '@user/SIGNED_OUT',
}

interface UserActionType<T, P> {
  type: T;
  payload: P;
}

type UserAction =
  | UserActionType<typeof EUserActions.SIGNED_IN, User>
  | UserActionType<typeof EUserActions.SIGNED_OUT, null>;

export type { User, UserState, UserAction };
export { EUserActions };
