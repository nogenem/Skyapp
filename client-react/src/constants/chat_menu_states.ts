export const MENU_STATES = {
  MAIN: 1,
  CHANGING_LANGUAGE: 2,
  CHANGING_USER_STATUS: 3,
  CHANGING_USER_THOUGHTS: 4,
} as const;
type TMenuStates = typeof MENU_STATES[keyof typeof MENU_STATES];

export type { TMenuStates };
