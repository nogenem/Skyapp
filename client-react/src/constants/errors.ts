export const INVALID_EMAIL = "Invalid email";
export const CANT_BE_BLANK = "Can't be blank";
export const PASSWORDS_MUST_MATCH = "Passwords must match";
export const fieldIsTooShort = (n: number) =>
  `This field must be at least ${n} chars long`;
