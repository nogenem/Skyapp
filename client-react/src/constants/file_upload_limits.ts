export default {
  fileSize: 5 * 1024 * 1024, // 5MB in bytes
  files: 5,
};
