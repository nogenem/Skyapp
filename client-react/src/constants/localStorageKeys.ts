export const LOCAL_STORAGE_TOKEN = 'skyapp-jwt';
export const LOCAL_STORAGE_THEME_MODE = 'skyapp-theme-mode';
