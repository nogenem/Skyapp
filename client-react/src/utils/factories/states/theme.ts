import { initialState } from '~/redux/theme/reducer';
import type { TThemeState, TThemeMode } from '~/redux/theme/types';

interface IOptions {
  useConstValues: boolean;
  useInitialState: boolean;
}

export default (
  override?: Partial<TThemeState>,
  options?: Partial<IOptions>,
): TThemeState => {
  let mode: TThemeMode | undefined =
    Math.random() < 0.2 ? undefined : Math.random() < 0.5 ? 'dark' : 'light';
  if (options?.useConstValues) mode = 'dark';

  return options?.useInitialState
    ? initialState
    : ({
        mode,
        ...(override || {}),
      } as TThemeState);
};

export type { IOptions };
