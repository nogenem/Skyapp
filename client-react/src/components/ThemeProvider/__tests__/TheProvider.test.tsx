import React from 'react';
import { Provider } from 'react-redux';

import { render } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { LOCAL_STORAGE_THEME_MODE } from '~/constants/localStorageKeys';
import type { IAppState } from '~/redux/store';
import { TThemeMode } from '~/redux/theme/types';

import { ThemeProvider, UnconnectedThemeProvider } from '../index';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const emptyState: Partial<IAppState> = {};
const renderWithRedux = (
  ui: React.ReactNode,
  initialState: Partial<IAppState> = emptyState,
) => {
  const store = mockStore(initialState);
  return {
    ...render(<Provider store={store}>{ui}</Provider>),
  };
};

const Children = () => <div />;

// TODO: Add better tests for this !?
describe('Connected ThemeProvider', () => {
  it('renders correctly', async () => {
    const { container } = renderWithRedux(
      <ThemeProvider>
        <Children />
      </ThemeProvider>,
    );

    expect(container).toMatchSnapshot();
  });
});

describe('Unconnected ThemeProvider', () => {
  it('Trys to switch the theme mode to `dark`', async () => {
    const currentThemeMode = 'light' as TThemeMode;
    const switchMode = jest.fn();

    localStorage.setItem(LOCAL_STORAGE_THEME_MODE, 'dark');
    render(
      <UnconnectedThemeProvider
        switchMode={switchMode}
        themeMode={currentThemeMode}
      >
        <Children />
      </UnconnectedThemeProvider>,
    );

    expect(switchMode).toHaveBeenCalledTimes(1);
    expect(switchMode).toHaveBeenCalledWith('dark');
  });
});
