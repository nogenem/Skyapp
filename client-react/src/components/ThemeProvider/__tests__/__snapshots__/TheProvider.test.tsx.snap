// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connected ThemeProvider renders correctly 1`] = `
<div>
  <div />
</div>
`;
